### PR TITLE
Rancher proxy uris

### PR DESCRIPTION
**Pull request by mlhnono68**
**What this PR does / why we need it**:
This PR relates to `virtctl` and its unability to use kubeconfigs provided by RANCHER. The underlying reason is that a RANCHER cluster URL will typically have a path describing the proxied cluster in the querystring (e.g. : `https://rancher-url/k8s/clusters/c-m-someid`) whereas Kubernetes cluster URLs typically have no path (e.g. : `https://192.168.3.3:6443
`)

As a consequence, using `virtctl` on RANCHER proxied clusters typically wrongly defines the URIs to use to call Kubernetes APIs. For example, calling `virtctl start xxx` on a RANCHER cluster would hit `https://rancher-url/apis/subresources.kubevirt.io/v1alpha3/namespaces/default/virtualmachines/xxx/start` instead of `https://rancher-url//k8s/clusters/c-m-someid/apis/subresources.kubevirt.io/v1alpha3/namespaces/default/virtualmachines/xxx/start`.

This pull requests aims at computing the URIs to use correctly so that `virtctl` can also work on RANCHER clusters.

It further extends fixing https://github.com/kubevirt/kubevirt/issues/3760

**Release note**:

```release-note
Fixes virtctl to support connection to clusters proxied by RANCHER or having special paths
```

### CODE REVIEW

File: staging/src/kubevirt.io/client-go/kubecli/version.go
Comment (vasiliy-ul) [inline staging/src/kubevirt.io/client-go/kubecli/version.go:60]: Ideally, `submatchall` needs to be null-checked. But maybe a simple `re.FindString` would be more appropriate here?

```go
sBasePath := re.FindString(v.restClient.Get().URL().Path)
```

File: staging/src/kubevirt.io/client-go/kubecli/vm.go
Comment (vasiliy-ul) [inline staging/src/kubevirt.io/client-go/kubecli/vm.go:45]: Same here.

File: staging/src/kubevirt.io/client-go/kubecli/version.go
Comment (vasiliy-ul) [inline staging/src/kubevirt.io/client-go/kubecli/version.go:58]: Hm... I think this may not work as expected if there is already an `apis` element in the path. What about instead parsing the URL from the config? Like its done in:

https://github.com/kubevirt/kubevirt/blob/74781324b726a6d58799be1257a34f61e6749513/staging/src/kubevirt.io/client-go/kubecli/vmi.go#L135-L141

**Review by vasiliy-ul — COMMENTED** [review]
Hey, nice fix. I added a couple of comments inline.

/ok-to-test

File: staging/src/kubevirt.io/client-go/kubecli/vm.go
Comment (mlhnono68) [reply staging/src/kubevirt.io/client-go/kubecli/vm.go]: Same fix

File: staging/src/kubevirt.io/client-go/kubecli/version.go
Comment (vasiliy-ul) [inline staging/src/kubevirt.io/client-go/kubecli/version.go:63]: Probably, we can make it even simpler:
```suggestion
	u, err := url.Parse(v.config.Host)
	if err != nil {
		return nil, err
	}
	uri := path.Join(u.Path, ApiGroupName)
```
I think this should work without `Replace` as `v.config.Host` contains the initial cluster URL.

Then further you can do:
```go
uri = fmt.Sprintf("%s/apis/%s/version", u.Path, group.PreferredVersion.GroupVersion)
```

**Review by vasiliy-ul — COMMENTED** [review]
One suggestion inline to simplify the approach. Also, could you please check your git history? I think there are redundant merge commits. Better to squash all in one.

File: staging/src/kubevirt.io/client-go/kubecli/version.go
Comment (mlhnono68) [reply staging/src/kubevirt.io/client-go/kubecli/version.go]: I see. I actually now also changed in on vm and vmi.go the same way. I had to add the v.config for version which was not existing yet

**Review by vasiliy-ul — COMMENTED** [review]
> I think we should be good to go now @vasiliy-ul

Almost :) The fix looks good from my perspective. However, you currently have an issue with formatting. You need to build the code locally. This will automatically run the script to re-format everything appropriately. Then you need to include those changes in the PR. Otherwise, CI will not pass.

One more thing: could you please squash everything into one commit and provide a descriptive message? You can just take the PR description since it's good enough.

UPD: and in fact you can also add a release note since it's a user-facing improvement.

**Comment by mlhnono68** [comment]
Thanks for your help  @vasiliy-ul . I added release notes, squashed into one commit and also re-formatted using a local build. I hope I have not missed anything more :)

**Comment by vasiliy-ul** [comment]
> Thanks for your help @vasiliy-ul . I added release notes, squashed into one commit and also re-formatted using a local build. I hope I have not missed anything more :)

Thank you, @mlhnono68, it looks good. Still: could you please add a message to the commit (you can use `git commit --amend`)? Keeping the history in SCM is also important. As I mentioned earlier, you can just use the PR description. E.g.
```
Support for RANCHER URIs

The patch relates to virtctl and its unability to use kubeconfigs
provided by RANCHER. The underlying reason is that a RANCHER cluster URL
will typically have a path describing the proxied cluster in the
querystring (e.g.: https://rancher-url/k8s/clusters/c-m-someid) whereas
Kubernetes cluster URLs typically have no path (e.g.: https://192.168.3.3:6443)

As a consequence, using virtctl on RANCHER proxied clusters typically
wrongly defines the URIs to use to call Kubernetes APIs. For example,
calling virtctl start xxx on a RANCHER cluster would hit

    https://rancher.dev1.anywhere.navify.com/apis/subresources.kubevirt.io/v1alpha3/namespaces/default/virtualmachines/xxx/start

instead of

    https://rancher-url//k8s/clusters/c-m-someid/apis/subresources.kubevirt.io/v1alpha3/namespaces/default/virtualmachines/xxx/start.

This patch aims at computing the URIs to use correctly so that virtctl
can also work on RANCHER clusters.

```

**Comment by mlhnono68** [comment]
Makes sense. I updated the commit message and removed also the left-over reference to some RANCHER environment I was using while uncovering the bug

**Review by vasiliy-ul — APPROVED** [review]
Awesome. Looks good now :+1:

File: staging/src/kubevirt.io/client-go/kubecli/vm.go
Comment (xpivarc) [inline staging/src/kubevirt.io/client-go/kubecli/vm.go:193]: Why does the `base` of `restClient` don't safe as here?

**Comment by xpivarc** [comment]
Looks good I have left one question. Other than that do we also want to modify `guestfs.go`?

File: staging/src/kubevirt.io/client-go/kubecli/vm.go
Comment (mlhnono68) [reply staging/src/kubevirt.io/client-go/kubecli/vm.go:193]: I don't understand your question @xpivarc sorry. Can you clarify ?

**Comment by mlhnono68** [comment]
Good point about `guestfs.go`. I now also added it to the merge request @xpivarc . I hope you're ok to approve it with this additional support.

**Review by vasiliy-ul — APPROVED** [review]
@mlhnono68, the failure of `pull-kubevirt-fossa` is not related to your PR. It's a known issue. The job is not mandatory, so it will not block the merge. No need to retest it.

Ping @xpivarc, maybe you could take one more look here?

File: staging/src/kubevirt.io/client-go/kubecli/vm.go
Comment (xpivarc) [reply staging/src/kubevirt.io/client-go/kubecli/vm.go:193]: I was looking into how the uri is constructed and I think the `base`(attribute of restClient) should cover this case and we would not need this special handling. I did not have time to check.
@vasiliy-ul wdyt?

Comment (vasiliy-ul) [reply staging/src/kubevirt.io/client-go/kubecli/vm.go:193]: Hard to tell by looking just at the code... But isn't it a private attribute?

Comment (xpivarc) [reply staging/src/kubevirt.io/client-go/kubecli/vm.go:193]: Yes. It is probably populated from kubeconfig. That's why I am not sure we need this workaround. We should have a look at it as it would be easier than keeping this patch.

Comment (vasiliy-ul) [reply staging/src/kubevirt.io/client-go/kubecli/vm.go:193]: Not sure if it can work with `base` from `RESTClient` since in the code we overwrite the URI for each request with `RequestURI(...)`. I guess this is needed to set the proper VM/VMI subresource path (which is not included in the `base` attribute as kubeconfig provides only the 'Host' of the base URI).

Comment (vasiliy-ul) [reply staging/src/kubevirt.io/client-go/kubecli/vm.go:193]: UPD: hm... but maybe you are right. if we drop `RequestURI(...)` and use smth like `SubResource(...)` instead this will probably work.

**Pull request merged**
